File: software/examples/csharp/ExampleLoopback.cs
using System;
using Tinkerforge;

class Example
{
	private static string HOST = "localhost";
	private static int PORT = 4223;
	private static string UID = "XYZ"; // Change XYZ to the UID of your CAN Bricklet

	// Callback function for frame read callback
	static void FrameReadCB(BrickletCAN sender, byte frameType, long identifier,
	                        byte[] data, byte length)
	{
		if(frameType == BrickletCAN.FRAME_TYPE_STANDARD_DATA)
		{
			Console.WriteLine("Frame Type: Standard Data");
		}
		else if(frameType == BrickletCAN.FRAME_TYPE_STANDARD_REMOTE)
		{
			Console.WriteLine("Frame Type: Standard Remote");
		}
		else if(frameType == BrickletCAN.FRAME_TYPE_EXTENDED_DATA)
		{
			Console.WriteLine("Frame Type: Extended Data");
		}
		else if(frameType == BrickletCAN.FRAME_TYPE_EXTENDED_REMOTE)
		{
			Console.WriteLine("Frame Type: Extended Remote");
		}

		Console.WriteLine("Identifier: " + identifier);
		Console.Write("Data (Length: " + length + "):");

		for (int i = 0; i < length && i < 8; ++i) {
			Console.Write(" " + data[i]);
		}

		Console.WriteLine("");
		Console.WriteLine("");
	}

	static void Main()
	{
		IPConnection ipcon = new IPConnection(); // Create IP connection
		BrickletCAN can = new BrickletCAN(UID, ipcon); // Create device object

		ipcon.Connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Configure transceiver for loopback mode
		can.SetConfiguration(BrickletCAN.BAUD_RATE_1000KBPS,
		                     BrickletCAN.TRANSCEIVER_MODE_LOOPBACK, 0);

		// Register frame read callback to function FrameReadCB
		can.FrameReadCallback += FrameReadCB;

		// Enable frame read callback
		can.EnableFrameReadCallback();

		// Write standard data frame with identifier 1742 and 3 bytes of data
		byte[] data = new byte[8]{42, 23, 17, 0, 0, 0, 0, 0};
		can.WriteFrame(BrickletCAN.FRAME_TYPE_STANDARD_DATA, 1742, data, 3);

		Console.WriteLine("Press enter to exit");
		Console.ReadLine();

		can.DisableFrameReadCallback();

		ipcon.Disconnect();
	}
}

File: software/examples/php/ExampleLoopback.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletCAN.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletCAN;

const HOST = 'localhost';
const PORT = 4223;
const UID = 'XYZ'; // Change XYZ to the UID of your CAN Bricklet

// Callback function for frame read callback
function cb_frameRead($frame_type, $identifier, $data, $length)
{
    echo "Frame Type: " . $frame_type . "\n";
    echo "Identifier: " . $identifier . "\n";
    echo "Data (Length: " . $length . "):";

    for ($i = 0; $i < $length && $i < 8; ++$i) {
        echo " " . $data[$i];
    }

    echo "\n";
    echo "\n";
}

$ipcon = new IPConnection(); // Create IP connection
$can = new BrickletCAN(UID, $ipcon); // Create device object

$ipcon->connect(HOST, PORT); // Connect to brickd
// Don't use device before ipcon is connected

// Configure transceiver for loopback mode
$can->setConfiguration(BrickletCAN::BAUD_RATE_1000KBPS,
                       BrickletCAN::TRANSCEIVER_MODE_LOOPBACK, 0);

// Register frame read callback to function cb_frameRead
$can->registerCallback(BrickletCAN::CALLBACK_FRAME_READ, 'cb_frameRead');

// Enable frame read callback
$can->enableFrameReadCallback();

// Write standard data frame with identifier 1742 and 3 bytes of data
$data = [42, 23, 17, 0, 0, 0, 0, 0];
$can->writeFrame(BrickletCAN::FRAME_TYPE_STANDARD_DATA, 1742, $data, 3);

echo "Press ctrl+c to exit\n";
$ipcon->dispatchCallbacks(-1); // Dispatch callbacks forever

?>
